### ISSUE TITLE
JWS ECDSA signatures should not be ASN.1 DER formatted

### ISSUE THREAD

**Issue by martinvw** [jwtk/jjwt#125]
After some security auditing it appeared that when using JJWT with elliptic curve, only a part of the signature is taken into account when validating. Than means than one can append any value to it, it could be the letter 'a' or even an XSS string which could reflect back...

I wrote a test to explain the problem, I can create a pull-request for it. But note that I only added the test, I did not yet resolve the problem.

The test can be found here: https://github.com/martinvw/jjwt/blob/master/src/test/groovy/io/jsonwebtoken/SignaturePostfix.groovy

**Comment by lhazlewood** [comment]
I found a way where we can test this accurately using RFC 7520 test vectors:

https://tools.ietf.org/html/rfc7520#section-4.3

The trick here is that ECDSA will automatically generate random bytes during signature calculation (without us controlling that) so we have to 'fake' it to ensure the random number generator returns the same data in the test vectors.

The `FixedSecureRandom` technique outlined in [this SO question](http://stackoverflow.com/questions/19904548/calculating-ecdsa-signature-in-java-as-per-an-rfc-test-vector?rq=1) looks like it could be the solution to that problem.

**Comment by lhazlewood** [comment]
After researching this more, I'm fairly confident this is **not** a security issue - just a compatibility one.  IIUC the ECDSA signature verification algorithm will only use the first N bytes of the specified signature and ignore any remaining.  So appending data to an ECDSA signature before validation isn't a valid test.

What *does* need to be done however is to ensure JWS-compatible byte array encoding for ECDSA signatures.

Currently JJWT returns the signature bytes computed directly by the Java JCA implementation, which according to JCA documentation are ASN.1 DER bytes.  See [Algorithm Specifications](http://docs.oracle.com/javase/8/docs/technotes/guides/security/StandardNames.html#algspec) --> **The Digital Signature Algorithm** table --> **Signature Algorithm** row

JWS ECDSA signatures however need to be formatted instead as the EC point R and S unsigned integers converted to byte arrays, concatenated.  From https://tools.ietf.org/html/rfc7515#page-45:

> The result of the digital signature is the Elliptic Curve (EC) point (R, S), where R and S are unsigned integers ... the R and S values ... [are] octet sequences representing big-endian integers ... The JWS Signature is the value R || S

In other words (pseudo code):

```
byte[] signature = R.toUnsignedInteger().toByteArray() + S.toUnsignedInteger().toByteArray()
```

So, while this isn't at a security risk, this just means that until we fix this, signatures created by JJWT may only be verified by JJWT and are not compatible with other JWT/JWS libraries.

I'll use this ticket to reflect the work to make ECDSA signatures formatted per the JWS spec.

**Comment by martintreurnicht** [comment]
I've created a pull request to fix this issue, see https://github.com/jwtk/jjwt/pull/137

**Comment by lhazlewood** [comment]
resolved via #137
